### PR TITLE
frame_fix : now transforms the tf and odom message to base_link

### PR DESCRIPTION
**Pull request by GeorgiosGkovo**

### CODE REVIEW

File: ros/ros1/OdometryServer.hpp [hunk 1]
```diff
@@ -70,6 +74,11 @@ class OdometryServer {
     /// Global/map coordinate frame.
     std::string odom_frame_{"odom"};
     std::string child_frame_{"base_link"};
+
+    //necessary for changing frames
+    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+    std::shared_ptr<tf2_ros::TransformListener> tf2_listener_;
+    geometry_msgs::TransformStamped os_to_base_link_tf_;
```
[the remark this answers is not in the document]
Comment (GeorgiosGkovo) [reply ros/ros1/OdometryServer.hpp:81]: changed to sensor_to_child_frame_msg_

File: ros/ros1/OdometryServer.cpp [hunk 2]
```diff
@@ -103,6 +103,16 @@ OdometryServer::OdometryServer(const ros::NodeHandle &nh, const ros::NodeHandle
         br.sendTransform(alias_transform_msg);
     }
 
+    try
+    {
+        tf_buffer_->canTransform("os_sensor", "base_link", ros::Time::now(), ros::Duration(5.0));
+        os_to_base_link_tf_ = tf_buffer_->lookupTransform("os_sensor", "base_link", ros::Time(0));
```
[the remark this answers is not in the document]
Comment (GeorgiosGkovo) [reply ros/ros1/OdometryServer.cpp:109]: added sensor_frame_ as a loaded param

File: ros/ros1/OdometryServer.cpp [hunk 3]
```diff
@@ -137,18 +148,32 @@ void OdometryServer::RegisterFrame(const sensor_msgs::PointCloud2::ConstPtr &msg
         transform_msg.transform.translation.x = t_current.x();
         transform_msg.transform.translation.y = t_current.y();
         transform_msg.transform.translation.z = t_current.z();
-        tf_broadcaster_.sendTransform(transform_msg);
+
+        tf2::Transform odom_to_sensor_tf;
+        tf2::fromMsg(transform_msg.transform, odom_to_sensor_tf);
+
+        tf2::Transform os_to_base_link_tf;
```
[the remark this answers is not in the document]
Comment (GeorgiosGkovo) [reply ros/ros1/OdometryServer.cpp:155]: changed

File: ros/ros1/OdometryServer.cpp [hunk 4]
```diff
@@ -137,18 +148,32 @@ void OdometryServer::RegisterFrame(const sensor_msgs::PointCloud2::ConstPtr &msg
...
+        tf2::fromMsg(os_to_base_link_tf_.transform, os_to_base_link_tf);
+
+        tf2::Transform result_transform_tf = odom_to_sensor_tf * os_to_base_link_tf;
+        result_transform.transform = tf2::toMsg(result_transform_tf);
```
[the remark this answers is not in the document]
Comment (GeorgiosGkovo) [reply ros/ros1/OdometryServer.cpp:159]: done

File: ros/ros1/OdometryServer.cpp [hunk 5]
```diff
@@ -125,6 +135,7 @@ void OdometryServer::RegisterFrame(const sensor_msgs::PointCloud2::ConstPtr &msg
     const Eigen::Quaterniond q_current = pose.unit_quaternion();
 
     // Broadcast the tf
+    geometry_msgs::TransformStamped result_transform ;
     if (publish_odom_tf_) {
         geometry_msgs::TransformStamped transform_msg;
```
[the remark this answers is not in the document]
Comment (GeorgiosGkovo) [reply ros/ros1/OdometryServer.cpp]: Yeah I saw this was unnecessary but I didn't wanna change the package too much cause there are tons of changes like that we could be making. Changed now.

**Review by liam-broek — APPROVED** [review]

File: ros/ros1/OdometryServer.cpp [hunk 6]
```diff
@@ -103,6 +104,16 @@ OdometryServer::OdometryServer(const ros::NodeHandle &nh, const ros::NodeHandle
...
     }
 
+    try
+    {
+        tf_buffer_->canTransform(sensor_frame_, child_frame_, ros::Time::now(), ros::Duration(5.0));
+        sensor_to_child_frame_msg = tf_buffer_->lookupTransform(sensor_frame_, child_frame_, ros::Time(0));
+    }
+    catch (tf2::TransformException& ex)
+    {
+        ROS_ERROR("Failed to wait for transform: %s", ex.what());
```
Comment (remco-r) [inline ros/ros1/OdometryServer.cpp:114]: shouldnt something happen here?

File: ros/ros1/OdometryServer.cpp [hunk 7]
```diff
@@ -125,30 +136,34 @@ void OdometryServer::RegisterFrame(const sensor_msgs::PointCloud2::ConstPtr &msg
...
+
+        tf2::Transform sensor_to_child_frame_tf;
+        tf2::fromMsg(sensor_to_child_frame_msg.transform, sensor_to_child_frame_tf);
+
+        odom_to_child_frame_msg.transform = tf2::toMsg(odom_to_sensor_tf * sensor_to_child_frame_tf);
+
+        odom_to_child_frame_msg.header.stamp = msg->header.stamp;
+        odom_to_child_frame_msg.header.frame_id = odom_frame_;
+        odom_to_child_frame_msg.child_frame_id = child_frame_;
+        odom_to_child_frame_msg.transform.translation.z = t_current.z();
```
[the remark this answers is not in the document]
Comment (GeorgiosGkovo) [reply ros/ros1/OdometryServer.cpp:154]: it doesn't anymore

File: ros/ros1/OdometryServer.cpp (hunk 6 above)
Comment (GeorgiosGkovo) [reply ros/ros1/OdometryServer.cpp]: throwing rutime_error now

**Pull request closed without merging**